Docs for Network migrations and Statesync (#102)
kwilteam/docs: docs/node/statesync.mdx

---
sidebar_position: 5
sidebar_label: "Statesync"
id: "Statesync"
title: "Using State Sync"
description: "Learn how to bootstrap a new node using statesync"
slug: /node/statesync
---

> A new node joining the network needs to catchup with the existing nodes to be able to participate in the network. The default approach the node uses to catch up with the network is to replay all the blocks from the genesis. This process can take a long time depending on the number of blocks in the chain. Statesync is a feature that allows a new node to bootstrap directly from a snapshot instead of replaying all the blocks from the genesis. This rapidly reduces the time it takes to sync a new node to the network, but the node will have a truncated history starting from the height of the snapshot.

Reviewer: "...rapidly reduces" describes a rate of change for reduction. Maybe just "greatly reduces"?

> A new node joining the network needs to catchup with the existing nodes to be able to participate in the network. The default approach the node uses to catch up with the network is to replay all the blocks from the genesis. This process can take a long time depending on the number of blocks in the chain. Statesync is a feature that allows a new node to bootstrap directly from a snapshot instead of replaying all the blocks from the genesis. This greatly reduces the time it takes to sync a new node to the network, but the node will have a truncated history starting from the height of the snapshot.

Reviewer 2: Grammar, typos, and wordiness:

```suggestion
Any new node joining a network needs to first catch up with the existing nodes. The default approach is to replay all the blocks from the genesis. Depending on the number of blocks in the chain, this process can take a long time.

Statesync is a feature that allows a new node to bootstrap directly from a snapshot instead of replaying all the blocks from the genesis. Statesync greatly reduces the time needed to sync a new node to a network; however, the node will have a truncated history starting from the height of the snapshot.
```

:::note
Statesync is not attempted on a partially synced node. If you have a partially synced node, it uses blocksync by default to sync the remaining blocks.

> :::

Reviewer 2: Wdym: "Is not attempted"? Does this mean: "State sync should not be used on a partially-synced node" ?

Author: Yeah

## Trusted Snapshot Providers

> To support statesync, each network should have atleast two trusted snapshot providers that are responsible for creating, distributing and validating snapshots. These trusted snapshot providers should have [snapshot creation](/docs/daemon/config/settings#appsnapshots) enabled and these provider's chain P2P and RPC endpoints should be accessible to the other nodes in the network to respond to the snapshot discovery and validation requests from the new nodes.

Reviewer: Is it possible to just run with 1? If they are trusted anyways, I don't see why 2 is necessary. I would imagine most users would just run two machines, which would seem redundant.

Author: umm, actually it's a constraint from comet, it won't start statesync if there are less than 2 providers.

> To support statesync, each network should have atleast two trusted snapshot providers that are responsible for creating, distributing and validating snapshots. These trusted snapshot providers should have [snapshot creation](/docs/daemon/config/settings#appsnapshots) enabled and these provider's chain P2P and RPC endpoints should be accessible to the other nodes in the network to respond to the snapshot discovery and validation requests from the new nodes.

Reviewer 2: Typos and grammar:

```suggestion
To support statesync, each network should have at least two trusted snapshot providers responsible for creating, distributing, and validating snapshots. These trusted snapshot providers should have [snapshot creation](/docs/daemon/config/settings#appsnapshots) enabled and their chain P2P and RPC endpoints accessible to other nodes in the network.
```

> Along with the trusted snapshot providers, other nodes in the network can also enable snapshots and distribute them to the joining nodes during the statesync process. However, a joining node only accepts these snapshots after validating them with the trusted snapshot providers.

Reviewer 2: Is this something that cometBFT does programatically, or is it a recommended best practice? If programatically, how do you set the "trusted provider"?

### Enabling Snapshot Creation
Snapshots can be enabled on the Kwil nodes by providing the following configuration in the `config.toml` file. Refer to the [snapshot](/docs/daemon/config/settings#appsnapshots) configuration for more details.

```toml
[app.snapshots]

# Enables snapshots
enabled = true

# Path to the snapshots directory
snapshot_dir = "snapshots"

# (Multiples of) Heights at which the database is snapshotted.
recurring_height = 10000

# Maximum number of snapshots to store on disk
max_snapshots = 3
```

> If snapshots are enabled, node starts to generate snapshots at every block height that is a multiple of `recurring_height`. The snapshots generated are stored on the disk at the location specified using the `snapshot_dir` configuration. These snapshots are the logical sql dumps of the underlying Postgres database, compressed and divided into chunks of 16MB each. 

Reviewer 2: ```suggestion
If snapshots are enabled, node generates snapshots at every block height that is a multiple of `recurring_height`. The snapshots generated are stored on the disk at the location specified using the `snapshot_dir` configuration. These snapshots are the logical sql dumps of the underlying Postgres database, compressed and divided into chunks of 16MB each. 
```

## Configuring Statesync

> The new node joining the network can be configured to use statesync by providing the following configuration in the `config.toml` file.

Reviewer 2: Remove passive voice:

```suggestion
Configure the new node to use statesync by providing the following configuration in the `config.toml` file.
```

# Snapshots are accepted for statesync only after verifying it with these trusted snapshot providers.
# Atleast 1 trusted rpc server is required for enabling state sync.
rpc_servers = "http://rpc1:26657,http://rpc2:26657"

# Time to spend discovering snapshots before initiating a restore. (default: 15 seconds)
discovery_time = "15s"

# The timeout duration before re-requesting a chunk, possibly from a different
# peer (default: 1 minute), if the current peer is unresponsive to the chunk request.
chunk_request_timeout = "10s"

# Note: If the requested chunk is not received within 2 minutes (hard-coded default), 
# the state sync process is aborted and the node will fall back to the regular block sync process.
```

When state sync is enabled, the node first discovers snapshots from all its connected peers. It then selects the latest snapshot from those discovered and validates the integrity of the snapshot with the trusted snapshot provider. Once a valid snapshot is identified, the node fetches the snapshot chunks from the peers and restores the database state using these chunks. The node then begins syncing blocks starting from the snapshot height.

:::note

> The node will stay in the discovery phase until a snapshot is discovered and validated. If there are no snapshots in the network, the node will be stuck in the discovery phase. To make progess, restart the node by disabling statesync to switch to using blocksync.

Reviewer 2: ```suggestion
The node will stay in the discovery phase until a snapshot is discovered and validated. If there are no snapshots in the network, the node will be stuck in the discovery phase. To make progess, disable statesync and restart the node. The node will then progress with blocksync.
```

Reviewer 2: Are there any docs we can hyperlink to on blocksync?

Author: Umm, I can add one. But there is not much to talk about it. I will probably add one at the top of this file.
